linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,516 @@
+    volatile_atomic int64_t insert_idx;
+    volatile_atomic int64_t consume_idx;
+    volatile_atomic int32_t task_array_size;
+    volatile_atomic int32_t pending_call_count;
+
+    SRW_LOCK_HANDLE srw_lock;
+    uint32_t list_index;
+
+    THREAD_HANDLE* thread_handle_array;
+} THREADPOOL;
+
+static void internal_close(THREADPOOL_HANDLE threadpool)
+{
+    bool should_wait_for_transition = false;
+    do
+    {
+        THREADPOOL_STATE current_state = interlocked_compare_exchange(&threadpool->state, THREADPOOL_STATE_CLOSING, THREADPOOL_STATE_OPEN);
+        if (current_state == THREADPOOL_STATE_NOT_OPEN || current_state == THREADPOOL_STATE_CLOSING)
+        {
+            should_wait_for_transition = false;
+        }
+        else if (current_state == THREADPOOL_STATE_OPENING)
+        {
+            should_wait_for_transition = true;
+        }
+        else
+        {
+            (void)interlocked_exchange(&threadpool->quitting, 1);
+
+            while (interlocked_add(&threadpool->pending_call_count, 0) != 0)
+            {
+                ThreadAPI_Sleep(1);
+            }
+
+            for (int32_t index = 0; index < threadpool->used_thread_count; index++)
+            {
+                int dont_care;
+                if (ThreadAPI_Join(threadpool->thread_handle_array[index], &dont_care) != THREADAPI_OK)
+                {
+                    LogError("Failure joining thread number %" PRId32 "", index);
+                }
+            }
+            (void)interlocked_exchange(&threadpool->state, THREADPOOL_STATE_NOT_OPEN);
+            should_wait_for_transition = true;
+        }
+    } while (should_wait_for_transition);
+}
+
+static int threadpool_work_func(void* param)
+{
+    if (param == NULL)
+    {
+        LogCritical("Invalid args: param: %p", param);
+    }
+    else
+    {
+        struct timespec ts;
+        THREADPOOL* threadpool = param;
+        int64_t current_index;
+
+        do
+        {
+            if (clock_gettime(CLOCK_REALTIME, &ts) == -1)
+            {
+                // Consider Log
+                LogError("Failure getting time from clock_gettime");
+            }
+            else
+            {
+                // Setup the timeout for the semaphore to 2 sec
+                ts.tv_sec += 2;
+                if (sem_timedwait(&threadpool->semaphore, &ts) != 0)
+                {
+                    // Timed out
+                }
+                else
+                {
+                    THREADPOOL_WORK_FUNCTION task_func = NULL;
+                    void* task_param = NULL;
+                    // Is this index waiting for us to complete it
+                    srw_lock_acquire_shared(threadpool->srw_lock);
+                    {
+                        int32_t existing_count = interlocked_add(&threadpool->task_array_size, 0);
+                        // We have a new item
+                        current_index = interlocked_increment_64(&threadpool->consume_idx) % existing_count;
```

Comment:
>urrent_index = interlocked_increment_64(&threadpool->consume_idx) % existing_count; [](http://example.com/codeflow?start=25&length=83)

I'm pretty sure we can skip over tasks as this is written.

If we increment this but we find a task in the WORKING or NOT_USED state, then no one will look at that task (until we wrap around in the array, which could be much later or never).

I don't think we can just increment this unless we know we are going to pull from the array.

So I think again the logic is:
1. check current index item with the interlocked compare exchange (which will sort of "lock" it to this thread if the exchange succeeds)
2. increment the index if we "won" and this thread "owns" the task.

Reply:
So we must rely on the semaphore then...I guess our other worker thread code I'm used to is safe to wake the workers extra times because they can no-op. But this will break the index counting if it does.

We should definitely fail on that log warning

I was thinking of this situation:
N=not used
WR=working
WA=waiting

consume_idx=0
[WA,N,N,N]

2 threads are pulling tasks

T1: increment consume_idx->1, get task[0] in WAITING and do the work
T2: increment consume_idx->2, get task[1] in NOT_USED

T1: increment consume_idx->3, get task[2] in NOT_USED

schedule another task
consume_idx=3
[N,WA,N,N]